vault_source.go

```diff
@@ -0,0 +1,171 @@
+	token         string
+	documentName  string
+}
+
+func FromVaultDocument(vault_token, hosts_address, document_name string) *JSONSource {
+	hostTools := HostTools{
+		winnerChannel:  make(chan string, 1),
+		errorChannel:   make(chan bool),
+		failureChannel: make(chan bool),
+
+		hostsAddress: hosts_address,
+		token:        vault_token,
+		documentName: document_name,
+	}
+
+	maxRetry := 2
+	for i := 0; i < maxRetry; i++ {
+		log.Println("[INFO] Finding best host...")
+		hostTools.findBestHost()
+		select {
+		case winner := <-hostTools.winnerChannel:
+			log.Println("[INFO] Winner:", winner)
+			document, err := hostTools.getVaultDocument(winner)
+			if err != nil {
+				log.Println("[ERROR] Vault document read error:", err)
+			}
+			if document != nil {
+				return &JSONSource{values: document}
+			}
+		case <-hostTools.failureChannel:
+			log.Println("[INFO] All host addresses have been checked, no winner found")
+		}
+	}
+	panic("Unable to procure vault document")
+}
+
+/////////////////////////////////////////
+
+func (this *HostTools) getVaultDocument(address string) (map[string]interface{}, error) {
+	response, err := this.vaultClient(address, this.documentName)
+	if err != nil {
+		return nil, err
+	}
+
+	document, err := parseDocument(response.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	return document.Data, nil
+}
+
+func parseDocument(responseBody io.Reader) (*SecretDocument, error) {
+	var document SecretDocument
+	decoder := json.NewDecoder(responseBody)
+	if err := decoder.Decode(&document); err != nil {
+		return nil, err
+	}
+	return &document, nil
+}
+
+/////////////////////////////////////////
+
+func (this *HostTools) findBestHost() {
+	go this.watchForProblems()
+	for _, host := range this.getHostList() {
+		go this.getVaultNode(host)
+	}
+}
+
+func (this *HostTools) vaultClient(address, path string) (*http.Response, error) {
+	client := &http.Client{}
+	request, err := http.NewRequest("GET", "http://"+address+":8200/v1/"+path, nil)
+	if err != nil {
+		return nil, err
+	}
+	request.Header.Add("X-Vault-Token", this.token)
+	response, err := client.Do(request)
+	if err != nil {
+		return nil, err
+	}
+
+	return response, nil
+}
+
+func (this *HostTools) getVaultNode(host string) {
+	response, err := this.vaultClient(host, "sys/health")
+	if err != nil {
+		log.Printf("[ERROR] Vault server is not accessible (%s): %s", host, err)
+	}
+	if response != nil {
+		switch response.StatusCode {
+		case 200:
+			this.winnerChannel <- host
+			return
+		case 429:
+			log.Printf("[ERROR] Vault server is unsealed but in standby mode (%s)", host)
+		case 500:
+			log.Printf("[ERROR] Vault server is sealed or not initialized (%s)", host)
+		}
+	}
+	this.errorChannel <- true
+}
+
+func (this *HostTools) getHostList() []string {
+	hosts, err := net.LookupHost(this.hostsAddress)
+	if err != nil {
+		log.Fatalf("[ERROR] %s", err)
+	}
+	this.numberOfHosts = len(hosts)
+	return hosts
+}
+
+func (this *HostTools) watchForProblems() {
```

Comment:
There will always be a `true` value given to `this.errorChannel` on each iteration. Can you see a way that this would not be the case?